File: backend/pkg/core/organization/service.go
package organization

import (
	"context"
	"encoding/json"
	"regexp"
	"strings"

	"backend/pkg/core"
	"backend/pkg/types"

	"github.com/google/uuid"
)

// Service provides business logic for organizations
type Service struct {
	orgStore     OrganizationStore
	accountStore AccountStore
}

// NewService creates a new organization service with the provided stores
func NewService(orgStore OrganizationStore, accountStore AccountStore) *Service {
	return &Service{
		orgStore:     orgStore,
		accountStore: accountStore,
	}
}

// GetByID retrieves an organization by its ID
func (s *Service) GetByID(ctx context.Context, id uuid.UUID) (*OrganizationResponse, error) {
	org, err := s.orgStore.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	return toResponse(org), nil
}

// GetBySlug retrieves an organization by its slug
func (s *Service) GetBySlug(ctx context.Context, slug string) (*OrganizationResponse, error) {
	org, err := s.orgStore.FindBySlug(ctx, slug)
	if err != nil {
		return nil, err
	}
	return toResponse(org), nil
}

// List retrieves all organizations
func (s *Service) List(ctx context.Context) ([]OrganizationResponse, error) {
	orgs, err := s.orgStore.List(ctx)
	if err != nil {
		return nil, err
	}

	result := make([]OrganizationResponse, len(orgs))
	for i, org := range orgs {
		result[i] = *toResponse(&org)
	}
	return result, nil
}

// Create creates a new organization
func (s *Service) Create(ctx context.Context, req CreateRequest) (*OrganizationResponse, error) {
	// Generate slug if not provided
	slug := req.Slug
	if slug == "" {
		slug = generateSlug(req.Name)
	}

	// Check if slug already exists
	existing, err := s.orgStore.FindBySlug(ctx, slug)
	if err != nil && err != core.ErrNotFound {
		return nil, err
	}
	if existing != nil {
		return nil, core.ErrAlreadyExists
	}

	// Convert social links
	var socialLinks map[string]interface{}
	if req.SocialLinks != nil {
		socialLinks = make(map[string]interface{})
		for k, v := range *req.SocialLinks {
			socialLinks[k] = v
		}
	}

	org := &types.Organization{
		ID:              uuid.New(),
		Name:            req.Name,
		Slug:            slug,
		Bio:             req.Bio,
		LogoURL:         req.LogoURL,
		WebsiteURL:      req.WebsiteURL,
		EmailPublic:     req.EmailPublic,
		SocialLinks:     socialLinks,
		MetaDescription: req.MetaDescription,
	}

	if err := s.orgStore.Save(ctx, org); err != nil {
		return nil, err
	}

	return toResponse(org), nil
}

// Update updates an existing organization
func (s *Service) Update(ctx context.Context, id uuid.UUID, req UpdateRequest) (*OrganizationResponse, error) {
	org, err := s.orgStore.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}

	// Check if new slug is unique
	if req.Slug != nil && *req.Slug != org.Slug {
		existing, err := s.orgStore.FindBySlug(ctx, *req.Slug)
		if err != nil && err != core.ErrNotFound {
			return nil, err
		}
		if existing != nil {
			return nil, core.ErrAlreadyExists
		}
		org.Slug = *req.Slug
	}

	// Apply updates
	if req.Name != nil {
		org.Name = *req.Name
	}
	if req.Bio != nil {
		org.Bio = req.Bio
	}
	if req.LogoURL != nil {
		org.LogoURL = req.LogoURL
	}
	if req.WebsiteURL != nil {
		org.WebsiteURL = req.WebsiteURL
	}
	if req.EmailPublic != nil {
		org.EmailPublic = req.EmailPublic
	}
	if req.SocialLinks != nil {
		socialLinks := make(map[string]interface{})
		for k, v := range *req.SocialLinks {
			socialLinks[k] = v
		}
		org.SocialLinks = socialLinks
	}
	if req.MetaDescription != nil {
		org.MetaDescription = req.MetaDescription
	}

	if err := s.orgStore.Update(ctx, org); err != nil {
		return nil, err
	}

	return toResponse(org), nil
}

// Delete removes an organization by its ID
func (s *Service) Delete(ctx context.Context, id uuid.UUID) error {
	return s.orgStore.Delete(ctx, id)
}

// JoinOrganization sets the organization for a user account
func (s *Service) JoinOrganization(ctx context.Context, accountID, orgID uuid.UUID) error {
	// Verify organization exists
	_, err := s.orgStore.FindByID(ctx, orgID)
	if err != nil {
		return err
	}

	// Get and update account
	account, err := s.accountStore.FindByID(ctx, accountID)
	if err != nil {
		return err
	}

	account.OrganizationID = &orgID
	return s.accountStore.Update(ctx, account)
}

// LeaveOrganization clears the organization for a user account
func (s *Service) LeaveOrganization(ctx context.Context, accountID uuid.UUID) error {
	account, err := s.accountStore.FindByID(ctx, accountID)
	if err != nil {
		return err
	}

	account.OrganizationID = nil
	return s.accountStore.Update(ctx, account)
}

// Helper functions

func generateSlug(name string) string {
	slug := strings.ToLower(name)
	slug = strings.ReplaceAll(slug, " ", "-")
	reg := regexp.MustCompile("[^a-z0-9-]")
	slug = reg.ReplaceAllString(slug, "")
	reg = regexp.MustCompile("-+")
	slug = reg.ReplaceAllString(slug, "-")
	slug = strings.Trim(slug, "-")
	return slug
}

func toResponse(org *types.Organization) *OrganizationResponse {
	socialLinks := make(map[string]string)
	if org.SocialLinks != nil {
		data, err := json.Marshal(org.SocialLinks)
		if err == nil {
			json.Unmarshal(data, &socialLinks)
		}
	}

	return &OrganizationResponse{
		ID:              org.ID,
		Name:            org.Name,
		Slug:            org.Slug,
		Bio:             stringValue(org.Bio),
		LogoURL:         stringValue(org.LogoURL),
		WebsiteURL:      stringValue(org.WebsiteURL),
		EmailPublic:     stringValue(org.EmailPublic),
		SocialLinks:     socialLinks,
		MetaDescription: stringValue(org.MetaDescription),
	}
}

func stringValue(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

File: backend/src/core/organization/service.rs
use std::{collections::BTreeMap, sync::Arc};

use serde_json::Value;
use uuid::Uuid;

use crate::error::AppError;

use super::{
    Organization, OrganizationAccountRepository, OrganizationCreateRequest, OrganizationRepository,
    OrganizationResponse, OrganizationUpdateRequest,
};

#[derive(Clone)]
pub struct OrganizationService {
    organizations: Arc<dyn OrganizationRepository>,
    accounts: Arc<dyn OrganizationAccountRepository>,
}

impl OrganizationService {
    pub fn new(
        organizations: Arc<dyn OrganizationRepository>,
        accounts: Arc<dyn OrganizationAccountRepository>,
    ) -> Self {
        Self {
            organizations,
            accounts,
        }
    }

    pub async fn get_by_id(&self, id: Uuid) -> Result<OrganizationResponse, AppError> {
        self.organizations.find_by_id(id).await.map(to_response)
    }

    pub async fn get_by_slug(&self, slug: &str) -> Result<OrganizationResponse, AppError> {
        self.organizations.find_by_slug(slug).await.map(to_response)
    }

    pub async fn list(&self) -> Result<Vec<OrganizationResponse>, AppError> {
        self.organizations
            .list()
            .await
            .map(|organizations| organizations.into_iter().map(to_response).collect())
    }

    pub async fn create(
        &self,
        request: OrganizationCreateRequest,
    ) -> Result<OrganizationResponse, AppError> {
        let slug = if request.slug.is_empty() {
            generate_slug(&request.name)
        } else {
            request.slug
        };
        match self.organizations.find_by_slug(&slug).await {
            Ok(_) => {
                return Err(AppError::Conflict("resource already exists".to_owned()));
            }
            Err(AppError::NotFound) => {}
            Err(error) => return Err(error),
        }

        let social_links = request.social_links.map(string_social_links);
        let mut organization = Organization {
            id: Uuid::new_v4(),
            name: request.name,
            slug,
            bio: request.bio,
            logo_url: request.logo_url,
            website_url: request.website_url,
            email_public: request.email_public,
            social_links,
            meta_description: request.meta_description,
            created_at: None,
            updated_at: None,
        };
        self.organizations.save(&mut organization).await?;
        Ok(to_response(organization))
    }

    pub async fn update(
        &self,
        id: Uuid,
        request: OrganizationUpdateRequest,
    ) -> Result<OrganizationResponse, AppError> {
        let mut organization = self.organizations.find_by_id(id).await?;
        if let Some(slug) = request.slug
            && slug != organization.slug
        {
            match self.organizations.find_by_slug(&slug).await {
                Ok(_) => {
                    return Err(AppError::Conflict("resource already exists".to_owned()));
                }
                Err(AppError::NotFound) => {}
                Err(error) => return Err(error),
            }
            organization.slug = slug;
        }
        if let Some(name) = request.name {
            organization.name = name;
        }
        if let Some(bio) = request.bio {
            organization.bio = Some(bio);
        }
        if let Some(logo_url) = request.logo_url {
            organization.logo_url = Some(logo_url);
        }
        if let Some(website_url) = request.website_url {
            organization.website_url = Some(website_url);
        }
        if let Some(email_public) = request.email_public {
            organization.email_public = Some(email_public);
        }
        if let Some(social_links) = request.social_links {
            organization.social_links = Some(string_social_links(social_links));
        }
        if let Some(meta_description) = request.meta_description {
            organization.meta_description = Some(meta_description);
        }
        self.organizations.update(&organization).await?;
        Ok(to_response(organization))
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        self.organizations.delete(id).await
    }

    pub async fn join_organization(
        &self,
        account_id: Uuid,
        organization_id: Uuid,
    ) -> Result<(), AppError> {
        self.organizations.find_by_id(organization_id).await?;
        if self
            .accounts
            .set_organization(account_id, Some(organization_id))
            .await?
        {
            Ok(())
        } else {
            Err(AppError::NotFound)
        }
    }

    pub async fn leave_organization(&self, account_id: Uuid) -> Result<(), AppError> {
        if self.accounts.set_organization(account_id, None).await? {
            Ok(())
        } else {
            Err(AppError::NotFound)
        }
    }
}

pub fn generate_slug(name: &str) -> String {
    let mut slug = String::with_capacity(name.len());
    for character in name.to_lowercase().chars() {
        let character = if character == ' ' { '-' } else { character };
        if (character.is_ascii_lowercase() || character.is_ascii_digit() || character == '-')
            && (character != '-' || !slug.ends_with('-'))
        {
            slug.push(character);
        }
    }
    slug.trim_matches('-').to_owned()
}

fn string_social_links(values: BTreeMap<String, String>) -> BTreeMap<String, Value> {
    values
        .into_iter()
        .map(|(key, value)| (key, Value::String(value)))
        .collect()
}

fn to_response(organization: Organization) -> OrganizationResponse {
    OrganizationResponse {
        id: organization.id,
        name: organization.name,
        slug: organization.slug,
        bio: organization.bio.unwrap_or_default(),
        logo_url: organization.logo_url.unwrap_or_default(),
        website_url: organization.website_url.unwrap_or_default(),
        email_public: organization.email_public.unwrap_or_default(),
        social_links: organization
            .social_links
            .unwrap_or_default()
            .into_iter()
            .filter_map(|(key, value)| value.as_str().map(|value| (key, value.to_owned())))
            .collect(),
        meta_description: organization.meta_description.unwrap_or_default(),
    }
}
